src/workerd/server/bundle-fs.c++

```diff
@@ -0,0 +1,433 @@
+#include "bundle-fs.h"
+
+namespace workerd::server {
+
+namespace {
+// A WorkerFileSystem directory delegate from the worker bundle
+// configuration. Each type of module in the bundle is represented as a file.
+// The directory structure is created based on the module names. For example,
+// if the bundle contains a module with the name "foo/bar/baz", it will be
+// represented as a directory "foo" with a subdirectory "bar" and a file
+// "baz" inside it. The directory structure and files are read-only. All
+// timestamps are set to the Unix epoch.
+class BundleDirectoryDelegate final: public workerd::DirectoryDelegate,
+                                     public kj::AtomicRefcounted {
+ public:
+  BundleDirectoryDelegate(config::Worker::Reader conf)
+      : inner(capnp::clone(conf)),
+        root(kj::atomicRefcounted<DirectoryEntry>()) {
+
+    kj::Path rootPath{};
+    DirectoryEntry* current = root.get();
+
+    switch (inner->which()) {
+      case config::Worker::Which::MODULES: {
+        for (auto module: inner->getModules()) {
```

Comment:
This will increase the startup time of all scripts, even those who might never us the filesystem.
We should set up lazy initialization for this type and make sure nothing in our codebase eagerly initializes the bundle filesystem unless it is actually used.
Either that or figure out a different data structure that lets us have an O(1) initialization).

Reply:
Given that this particular class is only used in workerd I'm not all that concerned with the additional cost. A separate version of this will need to be implemented in the internal repo using the different bundle configuration format. That one will have more performance optimizations.

Reply:
The internal and the external versions don't really need to be all that different, good code deduplication would have them be very thin wrappers around the same object.
I disagree with punting this question for later, a good implementation should have this thought through.

Reply:
While I generally agree, there is *nothing* in `workerd/server` that is reused in the internal repo. The two projects use two entirely separate configuration schemas. It's not about punting the question for later it's that I have yet to do that internal implementation. Once I do, there's an opportunity to separate out and deduplicate but right now there's nothing to deduplicate yet.